airflow/io/__init__.py

```diff
@@ -0,0 +1,297 @@
+        ...
+
+    @abstractmethod
+    def tell(self) -> int:
+        ...
+
+    @abstractmethod
+    def close(self) -> None:
+        ...
+
+    def __enter__(self) -> InputStream:
+        """Provide setup when opening an InputStream using a 'with' statement."""
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exctype: type[BaseException] | None,
+        excinst: BaseException | None,
+        exctb: TracebackType | None,
+    ) -> None:
+        """Perform cleanup when exiting the scope of a 'with' statement."""
+
+
+@runtime_checkable
+class OutputStream(Protocol):  # pragma: no cover
+    """A protocol for the file-like object returned by OutputFile.create(...).
+
+    This outlines the minimally required methods for a writable output stream returned from an OutputFile
+    implementation's `create(...)` method. These methods are a subset of IOBase/RawIOBase.
+    """
+
+    @abstractmethod
+    def write(self, b: bytes) -> int:
+        ...
+
+    @abstractmethod
+    def close(self) -> None:
+        ...
+
+    @abstractmethod
+    def __enter__(self) -> OutputStream:
+        """Provide setup when opening an OutputStream using a 'with' statement."""
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exctype: type[BaseException] | None,
+        excinst: BaseException | None,
+        exctb: TracebackType | None,
+    ) -> None:
+        """Perform cleanup when exiting the scope of a 'with' statement."""
+
+
+class InputFile(ABC):
+    """A base class for InputFile implementations.
+
+    Args:
+        location (str): A URI or a path to a local file.
+
+    Attributes:
+        location (str): The URI or path to a local file for an InputFile instance.
+        exists (bool): Whether the file exists or not.
+    """
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the input file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def open(self, seekable: bool = True) -> InputStream:
+        """Return an object that matches the InputStream protocol.
+
+        Args:
+            seekable: If the stream should support seek, or if it is consumed sequential.
+
+        Returns:
+            InputStream: An object that matches the InputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileNotFoundError: If the file at self.location does not exist.
+        """
+
+
+class OutputFile(ABC):
```

Comment:
I like the (basic) model but am not certain if it makes sense to separate the file / path like objects for reading and writing. Of course if I want to get an `InputStream` or `OutputStream` I need to have different IO classes but abstract wise the `File` Object is mainly the same irrespective if I want to open for reading or writing later?
Not sure about other opinions of if there is a functional dependency or use case but I'd rather merge `InputFile` and `OutputFile` together.

Reply:
The important part is to have a `file` like object so it can be used with, for example, `shutil.copyfileobj`. That might be covered by ObjectStoragePath, but I haven't checked yet.